spec/project-spec.coffee

```diff
@@ -197,6 +199,43 @@ describe "Project", ->
           atom.project.bufferForPath("b").then (anotherBuffer) ->
             expect(anotherBuffer).not.toBe buffer
 
+  describe ".repositoryForDirectory(directory)", ->
+    # it() causes beforeEach() in spec-helper.coffee to be run, which reassigns
+    # atom.project. Therefore, the atom.project used in waitsForPromise() will
+    # be different than the atom.project used in a subsequent it(). To work
+    # around this issue, we use runs() instead of it().
+
+    it "there is no repository for /tmp even though there are RepositoryProviders", ->
+      the_result = "dummy_value"
+
+      waitsForPromise ->
+        directory = new Directory("/tmp")
+        atom.project.repositoryForDirectory(directory).then (result) ->
+          the_result = result
+
+      runs ->
+        expect(atom.project.repositoryProviders.length).toBeGreaterThan 0
+        expect(the_result).toBe null
+        expect(atom.project.repositoryPromisesByPath.size).toBe 0
+
+    it "when Git repository for directory, promise resolves to Atom's GitRepository and is cached", ->
+      the_result = "dummy_value"
+      the_promise = null
+      directory = new Directory(path.join(__dirname, '..'))
+
+      waitsForPromise ->
+        the_promise = atom.project.repositoryForDirectory(directory)
+        the_promise.then (result) ->
+          the_result = result
+
+      runs ->
+        dirPath = directory.getRealPathSync()
+        expect(the_result).toBeInstanceOf GitRepository
+        expect(the_result.getPath()).toBe path.join(dirPath, '.git')
+
+        expect(atom.project.repositoryPromisesByPath.size).toBe 1
+        expect(atom.project.repositoryPromisesByPath.get dirPath).toBe the_promise
```

Comment:
I think that rather than using the private property `repositoryPromisesByPath`, you should verify the caching of the promises by saying:

```coffee
expect(atom.project.repositoryForDirectory(directory)).toBe(atom.project.repositoryForDirectory(directory))
```